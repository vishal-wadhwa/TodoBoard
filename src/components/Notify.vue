<template>
  <v-snackbar v-show='!!msg' v-model='snackbar' bottom auto-height>
    <v-icon :color='type' class='mr-3' medium>{{TYPE_TO_ICON[type]}}</v-icon>
    {{msg}}
    <v-btn color='pink' flat @click='snackbar = false'>Close</v-btn>
  </v-snackbar>
</template>
<script>
const TYPES = ['success', 'info', 'warning', 'error']

export default {
  props: {
    msg: String,
    type: {
      type: String,
      default: 'info',
      validator: v => TYPES.includes(v)
    }
  },
  data () {
    return {
      snackbar: false,
      TYPE_TO_ICON: {
        success: 'check_circle',
        info: 'info',
        warning: 'warning',
        error: 'error'
      }
    }
  },
  created () {
    this.snackbar = !!this.msg
  },
  watch: {
    msg (nv) {
      this.snackbar = true
    }
  }
}
</script>
