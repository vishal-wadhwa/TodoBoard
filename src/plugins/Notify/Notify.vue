<template>
  <transition name='v-snack-transition'>
    <div v-show='isActive' class='v-snack' :class='classes' :on='$listeners'>
      <div class='v-snack__wrapper' :style='setBackgroundColor(color, {}).style'>
        <div class='v-snack__content'>
          <v-icon :color='type' class='mr-3' medium>{{TYPE_TO_ICON[type]}}</v-icon>
          {{msg}}
          <v-btn color='pink' flat @click='onClose'>Close</v-btn>
        </div>
      </div>
    </div>
  </transition>
</template>
<script>
import { VSnackbar } from 'vuetify/lib'
import { EventBus } from './event_bus'

export default {
  name: 'Notify',
  extends: VSnackbar,
  data () {
    return {
      type: 'info',
      msg: '',
      TYPE_TO_ICON: {
        success: 'check_circle',
        info: 'info',
        warning: 'warning',
        error: 'error'
      }
    }
  },
  methods: {
    notify (data = {}) {
      if (!data.msg) return
      this.onClose()
      this.msg = data.msg
      this.type = data.type || 'info'
      // watch on this.isActive will trigger this.setTimeout()
      this.$nextTick(() => (this.isActive = true))
    },
    onClose () {
      this.isActive = false
      window.clearTimeout(this.activeTimeout)
    },
    setTimeout () {
      if (!this.isActive) return
      this.activeTimeout = window.setTimeout(this.onClose, this.timeout)
    }
  },
  created () {
    EventBus.$on('show', this.notify)
  }
}
</script>
