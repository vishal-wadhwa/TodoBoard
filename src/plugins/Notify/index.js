import Notify from './Notify'
import { EventBus } from './event_bus'

const Notif = {
  install (Vue, args = {}) {
    if (this.installed) {
      return
    }

    this.installed = true
    this.params = args

    Vue.component(args.componentName || 'notify', Notify)

    const notify = (params) => {
      if (typeof params === 'object') {
        EventBus.$emit('show', params)
      }
    }

    const name = args.name || 'notify'

    Vue.prototype['$' + name] = notify
    Vue[name] = notify
  }
}

export default Notif
