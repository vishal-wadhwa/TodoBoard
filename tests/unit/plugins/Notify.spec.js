import { shallowMount, createLocalVue, mount } from '@vue/test-utils'
import Notify from '@/plugins/Notify/Notify'
import Vuetify from 'vuetify'
import Vue from 'vue'
Vue.use(Vuetify)

describe('Notify.vue', () => {
  // required to vuetify [warn] in snapshot test
  // const app = document.createElement('div')
  // app.setAttribute('data-app', true)
  // document.body.appendChild(app)

  let localVue = null
  const factory = (opts = {}, full = false) => {
    return full ? mount(Notify, opts) : shallowMount(Notify, opts)
  }
  beforeEach(() => {
    localVue = createLocalVue()
    // localVue.use(Vuetify)
  })

  it('does not show notification if msg is empty', () => {
    const wrapper = factory({
      localVue
    })
    wrapper.vm.notify({ type: 'info', msg: '' })
    expect(wrapper.vm.isActive).toBeFalsy()
  })

  it('does not show notification if no data is provided', () => {
    const wrapper = factory({
      localVue
    })

    wrapper.vm.notify({ type: 'info', msg: '' })
    expect(wrapper.vm.isActive).toBeFalsy()
  })

  it('shows the notification with correct msg and type', async () => {
    const wrapper = factory({
      localVue
    })

    let msg = 'Notif 1'
    let type = 'warning'
    wrapper.vm.notify({ type, msg })
    await wrapper.vm.$nextTick()

    expect(wrapper.vm.msg).toEqual(msg)
    expect(wrapper.vm.type).toEqual(type)
    expect(wrapper.text()).toMatch(msg)
    expect(wrapper.text()).toMatch(type)
  })

  it('discards the previous notification and shows the new one immediately', async () => {
    const wrapper = factory({
      localVue
    })

    let msg = 'Notif 1'
    let type = 'warning'

    wrapper.vm.notify({ type, msg })
    await wrapper.vm.$nextTick()

    expect(wrapper.vm.msg).toEqual(msg)
    expect(wrapper.vm.type).toEqual(type)
    expect(wrapper.text()).toMatch(msg)
    expect(wrapper.text()).toMatch(type)

    msg = 'Notif 2'
    type = 'error'

    wrapper.vm.notify({ type, msg })
    await wrapper.vm.$nextTick()

    expect(wrapper.vm.msg).toEqual(msg)
    expect(wrapper.vm.type).toEqual(type)
    expect(wrapper.text()).toMatch(msg)
    expect(wrapper.text()).toMatch(type)
  })

  it('uses info icon by default', () => {
    const wrapper = factory({
      localVue
    })

    expect(wrapper.find('vicon-stub').text()).toMatch(wrapper.vm.TYPE_TO_ICON['info'])
  })

  it('shows the notification for 6000ms', async () => {
    jest.useFakeTimers()
    const wrapper = factory({
      localVue
    })
    let msg = 'Notif 1'
    let type = 'success'

    expect(wrapper.vm.isActive).toBeFalsy()

    wrapper.vm.notify({ type, msg })
    await wrapper.vm.$nextTick()

    expect(wrapper.vm.isActive).toBeTruthy()

    jest.runAllTimers()

    expect(wrapper.vm.isActive).toBeFalsy()
  })

  it('hides notification when Close button is clicked', async () => {
    const wrapper = factory({
      localVue
    }, true)
    let msg = 'Notif 1'
    let type = 'success'

    expect(wrapper.vm.isActive).toBeFalsy()

    wrapper.vm.notify({ type, msg })
    await wrapper.vm.$nextTick()

    expect(wrapper.vm.isActive).toBeTruthy()

    wrapper.find({ name: 'v-btn' }).trigger('click')
    await wrapper.vm.$nextTick()

    expect(wrapper.vm.isActive).toBeFalsy()
  })

  it('should match snapshot', () => {
    const wrapper = factory({
      localVue
    }, true)

    expect(wrapper.html()).toMatchSnapshot()
  })
})
