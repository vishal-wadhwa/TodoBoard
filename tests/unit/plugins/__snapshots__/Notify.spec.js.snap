// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notify.vue should match snapshot 1`] = `
<div on="[object Object]" class="v-snack v-snack--bottom" style="display: none;" name="v-snack-transition">
  <div class="v-snack__wrapper">
    <div class="v-snack__content"><i aria-hidden="true" class="v-icon mr-3 material-icons theme--light info--text" style="font-size: 28px;">info</i>

      <button type="button" class="v-btn v-btn--flat theme--light pink--text">
        <div class="v-btn__content">Close</div>
      </button></div>
  </div>
</div>
`;
